trollmoves/dispatcher.py

```diff
@@ -405,23 +285,114 @@ def create_dest_url(self, msg, client, conf):
+        # self.publisher.start()
+
+    def publish(self, msg, destinations, success):
+        """Publish a message.
+
+        The URI is replaced with the URI on the target server.
+        """
+        for url, _, client in destinations:
+            if not success[client]:
+                continue
             try:
-                self.publisher.stop()
-            except Exception:
-                logger.exception("Couldn't stop publisher.")
+                msg = self._get_new_message(msg, url, client)
+            except ValueError as err:
+                logger.error(str(err))
+                continue
+            logger.debug('Publishing %s', str(msg))
+            self.publisher.send(str(msg))
+
+    def _get_new_message(self, msg, url, client):
+        info = self._get_message_info(msg, url)
+        topic = self._get_topic(client, info)
+        if topic is None:
+            return None
+        return Message(topic, 'file', info)
+
+    def _get_message_info(self, msg, url):
+        info = msg.data.copy()
+        info["uri"] = urlsplit(url).path
+        return info
+
+    def _get_topic(self, client, info):
         try:
-            self.config_handler.close()
-        except Exception:
-            logger.exception("Couldn't stop config handler.")
+            topic = self.config[client]["publish_topic"]
+        except KeyError:
+            raise ValueError(f"Publish topic not configured for '{client}'")
+        return compose(topic, info)
+
+    def stop(self):
+        """Stop the reporter."""
+        self._pub_starter.stop()
+
+
+class PosttrollMessageIterator:
+    """Posttroll message iterator."""
+
+    def __init__(self, config):
+        """Set up the iterator."""
+        self.config = config
+        self.running = True
+
+    def __iter__(self):
+        """Iterate over messages from a listener container."""
+        with posttroll_listener(self.config) as listener:
+            while self.running:
+                try:
+                    yield listener.output_queue.get(timeout=0.05)
+                except Empty:
+                    continue
+
+    def stop(self):
+        """Stop the iterator."""
+        self.running = False
+
+
+@contextlib.contextmanager
+def posttroll_listener(new_config):
+    """Create a posttroll listener."""
+    subscriber_config = new_config.pop("posttroll_subscriber", {})
+    addresses = subscriber_config.pop('subscribe_addresses', None)
+    nameserver = subscriber_config.pop('nameserver', 'localhost')
+    services = subscriber_config.pop('subscribe_services', '')
+
+    topics = set()
+    for _client, client_config in new_config.items():
+        topics |= set(sum([item['topics'] for item in client_config['dispatch_configs']], []))
+
+    listener = ListenerContainer(topics=topics,
+                                 addresses=addresses,
+                                 nameserver=nameserver,
+                                 services=services)
+    yield listener
```

Comment:
Oooh, nice! This usage would have never crossed my mind.

Reply:
yes, context manager can be really nice sometimes :)